diff -u -r MHonArc-2.6.19.orig/lib/mhamain.pl MHonArc-2.6.19/lib/mhamain.pl
--- MHonArc-2.6.19.orig/lib/mhamain.pl	2014-04-21 22:33:10.000000000 -0400
+++ MHonArc-2.6.19/lib/mhamain.pl	2018-04-18 01:09:46.124640244 -0400
@@ -1562,7 +1562,7 @@
 ##
 sub defineIndex2MsgId {
     no warnings qw(deprecated);
-    if (!defined(%Index2MsgId)) {
+    if (!(%Index2MsgId)) {
 	foreach (keys %MsgId) {
 	    $Index2MsgId{$MsgId{$_}} = $_;
 	}
diff -u -r MHonArc-2.6.19.orig/lib/mhopt.pl MHonArc-2.6.19/lib/mhopt.pl
--- MHonArc-2.6.19.orig/lib/mhopt.pl	2012-01-29 14:45:55.000000000 -0500
+++ MHonArc-2.6.19/lib/mhopt.pl	2018-04-18 01:10:04.428560882 -0400
@@ -865,7 +865,7 @@
 sub update_data_2_1_to_later {
     no warnings qw(deprecated);
     # we can preserve filter arguments
-    if (defined(%main::MIMEFiltersArgs)) {
+    if (%main::MIMEFiltersArgs) {
 	warn qq/         preserving MIMEARGS...\n/;
 	%readmail::MIMEFiltersArgs = %main::MIMEFiltersArgs;
 	$IsDefault{'MIMEARGS'} = 0;
